src/Storage/Query/SelectQuery.php

```diff
@@ -0,0 +1,335 @@
+
+        $this->referenceJoins = [];
+        $this->fieldJoins = [];
+
+        foreach ($this->filters as $filter) {
+            if ($filter->getExpressionObject() instanceof \Doctrine\ORM\Query\Expr\Orx) {
+                // todo: `|||` and `bolt_field` integration.
+                $expr = $expr->add($filter->getExpression());
+            } elseif (in_array($filter->getKey(), $this->coreFields, true)) {
+                $expr = $expr->add($filter->getExpression());
+            } elseif (in_array($filter->getKey(), $this->referenceFields, true)) {
+                $this->referenceJoins[$filter->getKey()] = $filter;
+                $expr = $expr->add($filter->getExpression());
+            } else {
+                // This means the value is stored in the `bolt_field` table
+                $this->fieldJoins[$filter->getKey()] = $filter;
+            }
+        }
+
+        return $expr;
+    }
+
+    /**
+     * Returns all the parameters for the query.
+     */
+    public function getWhereParameters(): array
+    {
+        $params = [];
+        foreach ($this->filters as $filter) {
+            $params = array_merge($params, $filter->getParameters());
+        }
+
+        return $params;
+    }
+
+    /**
+     * Gets all the parameters for a specific field name.
+     *
+     * @param string $fieldName
+     *
+     * @return array array of key=>value parameters
+     */
+    public function getWhereParametersFor($fieldName): array
+    {
+        return array_intersect_key(
+            $this->getWhereParameters(),
+            array_flip(preg_grep('/^' . $fieldName . '_\d+$/', array_keys($this->getWhereParameters())))
+        );
+    }
+
+    /**
+     * Sets all the parameters for a specific field name.
+     *
+     * @param string $key
+     */
+    public function setWhereParameter($key, $value): void
+    {
+        foreach ($this->filters as $filter) {
+            if ($filter->hasParameter($key)) {
+                $filter->setParameter($key, $value);
+            }
+        }
+    }
+
+    public function addFilter(Filter $filter): void
+    {
+        $this->filters[] = $filter;
+    }
+
+    /**
+     * Returns all the filters attached to the query.
+     *
+     * @return Filter[]
+     */
+    public function getFilters(): array
+    {
+        return $this->filters;
+    }
+
+    /**
+     * Part of the QueryInterface this turns all the input into a Doctrine
+     * QueryBuilder object and is usually run just before query execution.
+     * That allows modifications to be made to any of the parameters up until
+     * query execution time.
+     */
+    public function build(): QueryBuilder
+    {
+        $query = $this->qb;
+        if ($this->getWhereExpression()) {
+            $query->where($this->getWhereExpression());
+        }
+        foreach ($this->getWhereParameters() as $key => $param) {
+            // xiao: `Connection::PARAM_STR_ARRAY` seems incorrect to me
```

Comment:
yes it is incorrect. use:
https://github.com/doctrine/doctrine2/blob/master/lib/Doctrine/ORM/Query/ParameterTypeInferer.php#L31